### PR TITLE
Refactor repro/minifier into CLI; add analyze

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100357
* #100354
* __->__ #100226

This is a two part PR; I can split it if you really want me to.

The first part is a refactor of the after aot repro/minifier scripts to come with a command line interface. I maintain exact BC with the previous interface (so, e.g., you still get a repro.py and a run_minifier.py that do the same thing as before), but each of these scripts also take command line arguments now which you can use to customize what actually happens. Check `run_repro` for full documentation on the arguments.

The second part of this is an implementation of `analyze` subcommand on the new CLI for any repro.

image

This facility is oriented towards accuracy debugging. It does several things:

1. It will run your model twice and check for nondeterminism in inductor/float64, *even* on intermediate inputs (our benchmarking nondeterminism test only checks for nondeterminism on the final output). This makes localizing which operator is nondeterministic easy.
2. It will run your compiled model side-by-side with eager and float64 variants, and then report when things diverge too far from RMSE delta from float64.

Importantly, it does all this without requiring every intermediate to be held in memory (which will cause an OOM on large repros, such as the one I tested this on.)

Some other minor improvements:

* MinifierTestBase now has an easy to comment out spot that you can use to retain the temporary directory; good for debugging
* We print "running minifier" and "running repro" in MinifierTestBase to make it easier to orient where logs are coming from
* same takes a `log_error` optional argument which you can use to reroute the error logs when things mismatch
* counters["inductor"]["intermediate_hooks"] tracks the number of intermediate hooks we've codegen'ed; good for populate the tqdm interface
* torch.fx.interpreter gets an official `boxed_run` interface which uses the boxed arguments calling convention and doesn't retain inputs unnecessarily long
* torch.utils._content_store gets compute_tensor_metadata/read_tensor_metadata helper functions for computing tensor information without serializing it

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by bertmaher — APPROVED** [review]
Very exciting, I can't wait to use this.

File: torch/_dynamo/repro/after_aot.py [hunk 1]
```diff
@@ -276,21 +287,27 @@ def symint(self, val):
...
             ContentStoreWriter(save_dir, stable_hash=stable_hash)
             if save_dir is not None
             else None
         )
         self.seen_storages = {}
 
+    def lines(self):
+        r = [
+            "import torch._dynamo.repro.after_aot",
+            f"reader = torch._dynamo.repro.after_aot.InputReader(save_dir={self.save_dir!r}, total={len(self._lines)})",
```
[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/_dynamo/repro/after_aot.py:306]: (btw this is very cool)

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/repro/after_aot.py:306]: whoopsie, you're right. Should count tensor too.

File: torch/_dynamo/repro/after_aot.py [hunk 2]
```diff
@@ -647,3 +601,277 @@ def inductor_accuracy_fails(fx_g, args, check_str=None):
...
+            diff_values = [(tuple1[i], tuple2[i]) for i in diff_indices]
+
+            if not diff_values:
+                return None
+            else:
+                return " and ".join(f"{a} != {b}" for a, b in diff_values)
+
+        def check_hook(name, val):
+            meta = writer.compute_tensor_metadata(val)
+            meta2 = reader.read_tensor_metadata(os.path.join("inductor", name))
```
Comment (bdhirsh) [inline torch/_dynamo/repro/after_aot.py:693]: Is OOM'ing due to reading the previous intermediate into memory while we're in the middle of execution a concern? We could probably make that less likely by just writing intermediates on both runs of the model, and then looping through the intermediates afterwards and reading them into memory + comparing them one at a time. That's definitely slower though, so I don't know if it's worth it.

File: torch/_dynamo/repro/after_aot.py [hunk 3]
```diff
@@ -647,3 +601,277 @@ def inductor_accuracy_fails(fx_g, args, check_str=None):
...
+                        tol=torch._dynamo.config.repro_tolerance,
+                        equal_nan=True,
+                        log_error=log_error,
+                    ):
+                        assert logged
+                    pbar.update(1)
+                return r
+
+        with tqdm(desc="Checking divergence", total=total) as pbar:
+            ReaderInterp(mod).boxed_run(args)
```
Comment (bdhirsh) [inline torch/_dynamo/repro/after_aot.py:778]: (Another OOM comment) - I suppose this could be run under FakeTensorMode too, so the only memory we use while interpreting the module is when we read into the memory the previously-saved inductor/eager intermediates.

**Review by bdhirsh — APPROVED** [review]
Left some comments mostly about reducing memory usage. This is really cool!

File: torch/_dynamo/repro/after_aot.py [hunk 4]
```diff
@@ -647,3 +601,277 @@ def inductor_accuracy_fails(fx_g, args, check_str=None):
...
+        # user specify the offload strategy.
+        from torch._inductor.hooks import intermediate_hook
+
+        with tqdm(desc="Compiling"):
+            compiled = compile_fx_inner(mod, args)
+        total = counters["inductor"]["intermediate_hooks"]
+
+        known_names = set()
+
+        def save_hook(name, val):
```
Comment (anijain2305) [inline torch/_dynamo/repro/after_aot.py:664]: Very very cool! I missed the last PR adding inductor hooks.

**Review by anijain2305 — APPROVED** [review]
This is amazing work. Use of Interpreter and hooks to check intermediates is amazing!

File: torch/_dynamo/repro/after_aot.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/repro/after_aot.py:693]: The current analysis memory usage is not optimal, but it is good enough for the particular case I was working on. My general attitude is that, if we discover we need the more memory efficient variants, we should implement them, but we always want to keep the faster more memory guzzling versions around, as they're much more pleasant when they work.

File: torch/_dynamo/repro/after_aot.py [hunk 5]
```diff
@@ -647,3 +601,277 @@ def inductor_accuracy_fails(fx_g, args, check_str=None):
...
+                name = n.name
+                if name in known_names:
+                    pbar.update(1)
+                    writer.write_tensor(os.path.join(self.subdir, name), r)
+                return r
+
+        # NB: the module cast doesn't actually do anything, since there are no
+        # parameters/buffers on the module
+        if not options.skip_saving_float64_intermediates:
+            new_mod, new_args = cast_to_fp64(mod, clone_inputs(args))
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/repro/after_aot.py:723]: So, while I was working on this PR, I noticed that we were doing a lot of unnecessary cloning in testing. I'd like to do something that's closer to tolerance (which will be faster), but that probably should be in its own diff. I'd prefer not to use the AOTAutograd analysis though: that analysis is fairly complicated, and when it comes to debugging, the simpler the better ;)

File: torch/_dynamo/repro/after_aot.py (hunk 3 above)
Comment (ezyang) [reply torch/_dynamo/repro/after_aot.py:778]: It can't actually, because we need the non-float64 eager intermediates

**Comment by ezyang** [comment]
Notes to reviewers: I did non-substantive changes to how the CLI interface works, changing the mutually exclusive `--analyze` flag to a subcommand so I can have different flags per subcommand. I also moved `minifier-query` into the script, deleting some last code from isolates; everything now goes through the CLI.
